lib/ide/src/controller/project.rs

```diff
@@ -78,47 +76,183 @@ impl Handle {
-        let transport           = MockTransport::new();
-        let project_ctrl        = controller::project::Handle::new(transport);
-        let file_manager_handle = project_ctrl.with_borrowed(|s| s.file_manager.clone());
-        let path                = Path("TestPath".to_string());
-        let another_path        = Path("TestPath2".to_string());
-
-        let text_ctrl         = project_ctrl.open_text_file(path.clone());
-        let same_text_ctrl    = project_ctrl.open_text_file(path.clone());
-        let another_text_ctrl = project_ctrl.open_text_file(another_path.clone());
-
-        assert!(file_manager_handle.identity_equals(&text_ctrl        .file_manager()));
-        assert!(file_manager_handle.identity_equals(&another_text_ctrl.file_manager()));
-        assert_eq!(path        , text_ctrl        .file_path()  );
-        assert_eq!(another_path, another_text_ctrl.file_path()  );
-        assert!(text_ctrl.identity_equals(&same_text_ctrl));
+        let mut executor    = LocalPool::new();
+        let finished        = Rc::new(RefCell::new(false));
+        let finished_clone  = finished.clone_ref();
+        let transport       = MockTransport::new();
+        set_spawner(executor.spawner());
+
+        spawn(async move {
+            let project_ctrl        = controller::project::Handle::new_running(transport);
+            let file_manager_handle = project_ctrl.file_manager();
+            let path                = Path("TestPath".to_string());
+            let another_path        = Path("TestPath2".to_string());
+
+            let text_ctrl        = project_ctrl.open_text_file(path.clone()).await.unwrap();
+            let same_text_ctrl   = project_ctrl.open_text_file(path.clone()).await.unwrap();
+            let another_txt_ctrl = project_ctrl.open_text_file(another_path.clone()).await.unwrap();
+
+            assert!(file_manager_handle.identity_equals(&text_ctrl       .file_manager()));
+            assert!(file_manager_handle.identity_equals(&another_txt_ctrl.file_manager()));
+            assert_eq!(path        , text_ctrl        .file_path()  );
+            assert_eq!(another_path, another_txt_ctrl.file_path()  );
+            assert!(text_ctrl.identity_equals(&same_text_ctrl));
+            *finished_clone.borrow_mut() = true;
+        });
+        executor.run_until_stalled();
+        assert!(*finished.borrow());
+    }
+
+    #[wasm_bindgen_test]
+    fn obtain_text_controller_for_module() {
+        let mut executor    = LocalPool::new();
+        let finished        = Rc::new(RefCell::new(false));
+        let finished_clone  = finished.clone_ref();
+        let mut transport   = MockTransport::new();
+        let transport_clone = transport.clone_ref();
+        set_spawner(executor.spawner());
+
+        spawn(async move {
+            let project_ctrl = controller::project::Handle::new_running(transport_clone);
+            let path         = Path(format!("test.{}", constants::LANGUAGE_FILE_EXTENSION));
```

Comment:
Why not use module::Location("Test".into()).to_path()?

We shouldn't hardcode provisional module->path logic in more than a single place.

Reply:
But this is test, isn't it? And we need anyway testing when a file is treated as module, and when not.